// ko/lib/actions/creator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackCreator = void 0;
const chalk_1 = __importDefault(require("chalk"));
const webpack_merge_1 = require("webpack-merge");
const webpack_1 = __importDefault(require("../webpack"));
const config_1 = __importDefault(require("../utils/config"));
class Creator {
}
class WebpackCreator extends Creator {
    constructor(opts) {
        super();
        this.opts = opts;
        this.baseConfig = this.initConfig(opts);
    }
    pluginsUnique(pluginNames) {
        return (0, webpack_merge_1.unique)('plugins', pluginNames, (plugin) => plugin.constructor && plugin.constructor.name);
    }
    initConfig(opts) {
        this.baseConfig = (0, webpack_1.default)(opts);
        return (0, webpack_merge_1.mergeWithCustomize)({
            customizeArray: this.pluginsUnique(['HtmlWebpackPlugin']),
        })(this.baseConfig, config_1.default.userConf);
    }
    mergeConfig(conf) {
        return (0, webpack_merge_1.merge)(conf);
    }
    successStdout(log) {
        console.log(chalk_1.default.green(log));
    }
    linkStdout(link) {
        console.log(chalk_1.default.underline(link));
    }
    errorStdout(log) {
        console.log(chalk_1.default.red(log));
    }
}
exports.WebpackCreator = WebpackCreator;
exports.default = Creator;
