// ko/lib/webpack/loaders/style.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const mini_css_extract_plugin_1 = require("mini-css-extract-plugin");
const autoprefixer_1 = __importDefault(require("autoprefixer"));
const CSS_LOADER = require.resolve('css-loader');
const LESS_LOADER = require.resolve('less-loader');
const SASS_LOADER = require.resolve('sass-loader');
const POSTCSS_LOADER = require.resolve('postcss-loader');
const antdV4Path = (0, path_1.join)(process.cwd(), 'node_modules/antd-v4');
const antdV4RealPath = (0, fs_1.existsSync)(antdV4Path)
    ? (0, fs_1.realpathSync)(antdV4Path)
    : antdV4Path;
const styleLoader = {
    loader: mini_css_extract_plugin_1.loader,
};
const cssLoader = {
    loader: CSS_LOADER,
    options: {
        sourceMap: true,
    },
};
//TODO: check postcss-loader should use sourceMap option
const postcssLoader = {
    loader: POSTCSS_LOADER,
    options: {
        sourceMap: true,
        postcssOptions: {
            plugins: [(0, autoprefixer_1.default)()],
        },
    },
};
const styleLoaders = [
    {
        test: /\.css$/,
        use: [styleLoader, cssLoader, postcssLoader],
    },
    {
        test: /\.s[ac]ss$/,
        use: [
            styleLoader,
            cssLoader,
            postcssLoader,
            {
                loader: SASS_LOADER,
                options: {
                    sourceMap: true,
                },
            },
        ],
    },
    {
        test: /\.less$/,
        exclude: [antdV4RealPath],
        use: [
            styleLoader,
            cssLoader,
            postcssLoader,
            {
                loader: LESS_LOADER,
                options: {
                    lessOptions: {
                        javascriptEnabled: true,
                    },
                    sourceMap: true,
                },
            },
        ],
    },
    {
        test: /\.less$/,
        include: [antdV4RealPath],
        use: [
            styleLoader,
            cssLoader,
            postcssLoader,
            {
                loader: LESS_LOADER,
                options: {
                    sourceMap: true,
                    lessOptions: {
                        modifyVars: {
                            '@ant-prefix': 'ant-v4',
                            '@font-size-base': '12px',
                            '@border-color-base': '#ddd',
                            '@font-family': "-apple-system, BlinkMacSystemFont, 'Segoe UI', 'PingFang SC', 'Hiragino Sans GB', 'Microsoft YaHei', 'Helvetica Neue', Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
                        },
                        javascriptEnabled: true,
                    },
                },
            },
        ],
    },
];
exports.default = styleLoaders;
